components/backup/src/writer.rs

```diff
@@ -0,0 +1,207 @@
+                    self.write.put(&data_key_write, &write.1)?;
+                    self.write_written = true;
+                }
+                TxnEntry::Prewrite { .. } => {
+                    return Err(Error::Other("prewrite is not supported".into()))
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Save buffered SST files to the given external storage.
+    pub fn save(mut self, storage: &dyn ExternalStorage) -> Result<Vec<File>> {
+        let name = self.name;
+        let save_and_build_file = |cf, mut contents: &[u8]| -> Result<File> {
+            BACKUP_RANGE_SIZE_HISTOGRAM_VEC
+                .with_label_values(&[cf])
+                .observe(contents.len() as _);
+            let name = format!("{}_{}.sst", name, cf);
+            let checksum = tikv_util::file::calc_crc32_bytes(contents);
+            storage.write(&name, &mut contents as &mut dyn std::io::Read)?;
+            let mut file = File::new();
+            file.set_crc32(checksum);
+            file.set_name(name);
+            Ok(file)
+        };
+        let start = Instant::now();
+        let mut files = Vec::with_capacity(2);
+        let mut buf = Vec::new();
+        if self.default_written {
+            // Save default cf contents.
+            buf.reserve(self.default.file_size() as _);
+            self.default.finish_into(&mut buf)?;
+            let default = save_and_build_file(CF_DEFAULT, &mut buf)?;
+            files.push(default);
+            buf.clear();
+        }
+        if self.write_written {
+            // Save write cf contents.
+            buf.reserve(self.write.file_size() as _);
+            self.write.finish_into(&mut buf)?;
+            let write = save_and_build_file(CF_WRITE, &mut buf)?;
+            files.push(write);
+        }
+        BACKUP_RANGE_HISTOGRAM_VEC
+            .with_label_values(&["save"])
+            .observe(start.elapsed().as_secs_f64());
+        Ok(files)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::path::Path;
+    use tempfile::TempDir;
+    use tikv::storage::TestEngineBuilder;
+
+    fn check_sst(ssts: &[(engine::CfName, &Path)], kvs: &[(engine::CfName, &[u8], &[u8])]) {
+        let temp = TempDir::new().unwrap();
+        let rocks = TestEngineBuilder::new()
+            .path(temp.path())
+            .cfs(&[engine::CF_DEFAULT, engine::CF_LOCK, engine::CF_WRITE])
+            .build()
+            .unwrap();
+        let db = rocks.get_rocksdb();
+
+        let opt = engine::rocks::IngestExternalFileOptions::new();
+        for (cf, sst) in ssts {
+            let handle = db.cf_handle(cf).unwrap();
+            db.ingest_external_file_cf(handle, &opt, &[sst.to_str().unwrap()])
+                .unwrap();
+        }
+        for (cf, k, v) in kvs {
+            let handle = db.cf_handle(cf).unwrap();
+            assert_eq!(*v, &*db.get_cf(handle, k).unwrap().unwrap());
+        }
```

Comment:
I think it's better if you also assert that there are no more keys in the rocksdb that doesn't exist in the `kvs`. You can scan all data from the rocksdb and compare with the `kvs`.